src/dir.c

```diff
@@ -1212,52 +1226,131 @@ check_tablespace_mapping(pgBackup *backup, bool incremental, bool *tblspaces_are
 			elog(ERROR, "--tablespace-mapping option's old directory "
 				 "doesn't have an entry in tablespace_map file: \"%s\"",
 				 cell->old_dir);
-
-		/* For incremental restore, check that new directory is empty */
-//		if (incremental)
-//		{
-//			if (!is_absolute_path(cell->new_dir))
-//				elog(ERROR, "tablespace directory is not an absolute path: %s\n",
-//					 cell->new_dir);
-//
-//			if (!dir_is_empty(cell->new_dir, FIO_DB_HOST))
-//				elog(ERROR, "restore tablespace destination is not empty: \"%s\"",
-//					 cell->new_dir);
-//		}
 	}
 
+	/*
+	 * There is difference between incremental restore of already existing
+	 * tablespaceses and remapped tablespaceses.
+	 * Former are allowed to be not empty, because we treat them like an
+	 * extension of PGDATA.
+	 * The latter are not, unless "--force" flag is used.
+	 * in which case the remapped directory is nuked - just to be safe,
+	 * because it is hard to be sure that there are no some tricky corner
+	 * cases of pages from different systems having the same crc.
+	 * This is a strict approach.
+	 *
+	 * Why can`t we not nuke it and just let it roll ?
+	 * What if user just wants to rerun failed restore with the same
+	 * parameters? Nuking is bad for this case.
+	 *
+	 * Consider the example of existing PGDATA:
+	 * ....
+	 * 	pg_tablespace
+	 * 		100500-> /somedirectory
+	 * ....
+	 *
+	 * We want to remap it during restore like that:
+	 * ....
+	 * 	pg_tablespace
+	 * 		100500-> /somedirectory1
+	 * ....
+	 *
+	 * Usually it is required for "/somedirectory1" to be empty, but
+	 * in case of incremental restore with 'force' flag, which required
+	 * of us to drop already existing content of "/somedirectory1".
+	 *
+	 * TODO: Ideally in case of incremental restore we must also
+	 * drop the "/somedirectory" directory first, but currently
+	 * we don`t do that.
+	 */
+
 	/* 2 - all linked directories must be empty */
 	for (i = 0; i < parray_num(links); i++)
 	{
 		pgFile	   *link = (pgFile *) parray_get(links, i);
 		const char *linked_path = link->linked;
 		TablespaceListCell *cell;
+		bool remapped = false;
 
 		for (cell = tablespace_dirs.head; cell; cell = cell->next)
 			if (strcmp(link->linked, cell->old_dir) == 0)
 			{
 				linked_path = cell->new_dir;
+				remapped = true;
 				break;
 			}
 
 		if (!is_absolute_path(linked_path))
-			elog(ERROR, "tablespace directory is not an absolute path: %s\n",
+			elog(ERROR, "Tablespace directory is not an absolute path: %s\n",
 				 linked_path);
 
 		if (!dir_is_empty(linked_path, FIO_DB_HOST))
 		{
+
 			if (!incremental)
```

Comment:
Why doesn't this new logic apply to full backups?

Reply:
It does, `incremental` flag here stands for `incremental restore`.